tools/tungsten/core/src/com/google/android/filament/tungsten/model/Nodes.kt

```diff
@@ -0,0 +1,106 @@
+ * limitations under the License.
+ */
+
+package com.google.android.filament.tungsten.model
+
+import com.google.android.filament.tungsten.compiler.GraphCompiler
+import com.google.android.filament.tungsten.compiler.Expression
+
+private val adderNodeCompile = fun(node: Node, compiler: GraphCompiler): Node {
+    val outputSlot = node.getOutputSlot("result")
+
+    val a = compiler.compileAndRetrieveVariable(node.getInputSlot("a"))
+    val b = compiler.compileAndRetrieveVariable(node.getInputSlot("b"))
+
+    val aExpression = a?.symbol ?: "float3(0.0f, 0.0f, 0.0f)"
+    val bExpression = b?.symbol ?: "float3(0.0f, 0.0f, 0.0f)"
+
+    val temp = compiler.getNewTemporaryVariableName("adder")
+    compiler.addCodeToMaterialFunctionBody("float3 $temp = $aExpression + $bExpression;\n")
+
+    compiler.setExpressionForOutputSlot(outputSlot, Expression(temp))
+
+    return node
+}
+
+private val shaderNodeCompile = fun(node: Node, compiler: GraphCompiler): Node {
+    val baseColor = compiler.compileAndRetrieveVariable(node.getInputSlot("baseColor"))
+    val metallic = compiler.compileAndRetrieveVariable(node.getInputSlot("metallic"))
+    val roughness = compiler.compileAndRetrieveVariable(node.getInputSlot("roughness"))
+
+    if (baseColor != null) {
+        compiler.addCodeToMaterialFunctionBody("material.baseColor.rgb = ${baseColor.symbol};\n")
+    }
+
+    if (metallic != null) {
+        compiler.addCodeToMaterialFunctionBody("material.metallic = ${metallic.symbol};\n")
+    }
+
+    if (roughness != null) {
+        compiler.addCodeToMaterialFunctionBody("material.roughness = ${roughness.symbol};\n")
+    }
+
+    return node
+}
+
+private val constantFloat3NodeCompile = fun(node: Node, compiler: GraphCompiler): Node {
+    val outputSlot = node.getOutputSlot("result")
+
+    val color = (node.properties[0].value as Float3)
+
+    val outputVariable = compiler.getNewTemporaryVariableName("float3Constant")
+    compiler.addCodeToMaterialFunctionBody(
+            "float3 $outputVariable = float3(${color.x}, ${color.y}, ${color.z});\n")
+
+    compiler.setExpressionForOutputSlot(outputSlot, Expression(outputVariable))
+    return node
+}
+
+val createAdderNode = fun(id: NodeId): Node {
+    return Node(
+        id = id,
+        type = "adder",
+        compileFunction = adderNodeCompile,
+        inputSlots = listOf("a", "b"),
+        outputSlots = listOf("result")
+    )
+}
+
+val createFloat3ConstantNode = fun(id: NodeId): Node {
+    return Node(
+        id = id,
+        type = "float3Constant",
+        compileFunction = constantFloat3NodeCompile,
+        outputSlots = listOf("result"),
+        properties = listOf(Property("value", Float3()))
+    )
+}
+
+val createShaderNode = fun(id: NodeId): Node {
+    return Node(
+        id = id,
+        type = "shader",
+        compileFunction = shaderNodeCompile,
+        inputSlots = listOf("baseColor", "metallic", "roughness"))
```

Comment:
You're missing a few input slots :p

Reply:
I sure am :) Just doing these as placeholders for now, will update to the full material model in the future